petrisrvn/petrisrvn.1
.TH PETRISRVN 1 "15 July 2021"
.\" $Header$
.SH NAME
petrisrvn \- solve stochastic rendezvous networks using GreatSPN.
.SH SYNOPSIS
.sp
.B petrisrvn
[
.B \-dknprvV
] [
.B \-m \fIfile\fR
] [
.B \-o \fIfile\fR
] [
.B \-P \fIpragma\fR
] [
.B \-z \fIspecialopts\fR
] [
filename
\&.\|.\|.
]
.SH DESCRIPTION
\fBPetrisrvn\fR is used to analytically solve stochastice rendezvous networks
using GreatSPN.  Refer to \fI``The Stochastic
Rendezvous Network Model for Performance of Synchronous Multi-tasking
Distributed Software''\fR for details of the model.  Refer to
\fI``SRVN Input File Format''\fR for a complete description of the
input file format for the programs.  See below for program
restrictions and limits.
.PP
\fBPetrisrvn\fR reads its input from \fIfilename\fR, specified at the
command line if present, or from the standard input otherwise.  By
default, output for an input file \fIfilename\fR specified on the
command line will be placed in the file \fIfilename\fB.out\fR.  If the
\fB\-p\fP switch is used, parseable output will also be written into
\fIfilename\fB.p\fR. If XML input is used, results will be written
back to the original input file.  This behaviour can be changed using
the \fB\-o\fIoutput\fR switch, described below.  If several files are
named, then each is treated as a separate model and output will be
placed in separate output files.  If input is from the standard input,
output will be directed to the standard output.  The file name `\-' is
used to specify standard input.
.PP
The \fB\-o\fIoutput\fR option can be used to direct output to the file
\fIoutput\fR regardless of the source of input.  Output will be XML if
XML input is used, parseable output if the \fB\-p\fP switch is used,
and normal output otherwise.  Multiple input files cannot be specified
when using this option.  Output can be directed to standard output by
using \fB\-o\fI\-\fR (i.e., the output file name is `\fI\-\fR'.)
.SH "OPTIONS"
.TP 
\fB\-d\fR, \fB\-\-debug\fR
Turn on debugging output.  Dump task and entry information showing
internal index numbers.  This option is useful for determining the
names of the servers and tasks since the parasol output routines do no
emit this information at present.  Output is directed to stdout unless
redirected using \fB\-m \fIfile\fR.
.TP
\fB\-I\fP, \fB\-\-input-format\fR=\fIarg\fR
This option is used to force the input file format to either \fIxml\fP or \fIlqn\fP.
By default, if the suffix of the input filename is one of: \fI.in\fP, \fI.lqn\fP or \fI.xlqn\fP, 
then the LQN parser will be used.  Otherwise, input is assumed to be XML.
.TP
\fB\-j\fR, \fB\-\-json\fR
Output results in JSON format.
.TP
\fB\-k\fR, \fB\-\-keep-net\fR
Keep all results from the solution of the petrinet.
.TP
\fB\-m\fR, \fB\-\-monitor-file\fR=\fI<filename>\fR
Save the output from the GreatSPN solver to \fIfilename\fP.  By
default, this output is discarded.
.TP
\fB\-n\fR, \fB\-\-no-execute\fR
Read the input, but do not solve.  No output is generated.
.TP
.TP
\fB\-o\fP, \fB\-\-output\fR=\fIfilename\fR
Direct analysis results to \fIfilename\fP.  A filename of `\fI-\fR'
directs output to standard output.  If \fIoutput\fR is a directory,
all output is saved in \fIoutput/input.out\fR. If the input model
contains a SPEX program with loops, the SPEX output is sent to
\fIoutput\fR; the individual model output files are found in the
directory \fIoutput.d\fR. If \fBpetrisrvn\fP is invoked with this 
option, only one input file can be specified.
.TP
\fB\-p\fR
Generate parseable output suitable as input to other programs such as
\fBsrvndiff(1)\fR and \fBlqn2ps(1)\fR. The parseable output is sent to an
output file, one for each input file specified, determined from its
corresponding input file by the addition of \fB.p\fR.  If standard
input is used as the input source, then the parseable output is output
to the standard output device.  In this case, only parseable output is
emitted.  If the \fB-o\fR \fIoutput\fR option is used, the parseable
output is sent to a file whose name is derived from \fIoutput\fR by
the addition of the suffix \fB.p\fR.  If a suffix already exists on
\fIoutput\fR then that suffix is replaced.
.TP
\fB\-P\fP, \fB\-\-pragma\fR=\fIvalue\fR
Set \fIpragma\fP to value.  See the section on pragmas below.
.TP
\fB\-Q\fP, \fB\-\-queue-size\fR=\fIn\fP
Set the queue size for open arrival or asynchronous sends to
\fIarg\fP.  Queue overflows are reported as lost messages.
.TP
\fB\-r\fP, \fB\-\-rtf\fR
Output results using Rich Text Format instead of plain text.
Processors, entries and tasks with high utilizations are coloured in
red.
.TP
\fB\-R\fR, \fB\-\-reload-net\fR
Reread a preexisting, solved model (generated using \fB\-k\fR).
.TP
\fB\-t\fR, \fB\-\-trace-greatspn\fR
Show output from the execution of the GreatSPN solver.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Display solution execution.  For \fBpetrisrvn\fP, the
output of the various petrinet solution stages is shown.
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version number.  
.TP
\fB\-w\fR, \fB\-\-no-warnings\fR
Ignore warnings.  The default is to print out all warnings.
.TP
\fB\-x\fR, \fB\-\-xml\fR
Output results in XML format.
.TP
\fB\-\-disjoint\-customers\fR
Create copies for reference tasks, rather that using a single task
with multiple tokens. (increases state space)
.TP
\fB\-\-no\-header\fR
Do not print out the Result Variable header when running with SPEX input.
This option has no effect otherwise.
.TP
\fB\-\-overtaking[=ARG]\fR
Find in-service and overtaking probabilities (increases state
space). ARG=task.
.TP
\fB\-\-queue\-limit=ARG\fR
Set the queue size at a task for any open arrivals or asynchronous
sends to ARG.  This amount is in addition to the possible of customers
that can arrive through rendezvous requests.
.TP
\fB\-\-random\-processors\fR
Use random queueing at all processors (reduces state space).
.TP
\fB\-\-random\-queueing\fR
Use random queueing at all tasks and processors (reduces state space).
.TP
\fB\-\-reload-lqx\fR
Re-run the LQX/SPEX program without re-solving the models.  Results must exist from a previous solution run.
This option is useful if LQX print statements or SPEX results are changed.
.TP
\fB\-\-restart\fR
Re-run the LQX/SPEX program without re-solving models which were
solved successfully.  Models which were not solved successfully because of state
space explosion, will be solved again.
This option is useful for running a second pass with different options
to reduce the state space.
.TP
\fB\-\-simplify\fR
Remove redundant parts of the net such as single place processors.
.TP
\fB\-\-spacing=ARG\fR
Set the spacing between places and transition to ARG for the GreatSPN editor (the default is
3).  
\fB\-\-debug-lqx\fR
Output debugging information as an LQX program is being parsed.
.TP
\fB\-\-debug-xml\fR
Output XML elements and attributes as they are being parsed.   Since the XML parser usually stops when it encounters an error,
this option can be used to localize the error.
.SH "PRAGMAS"
.TP
\fBforce-random-queueing\fR=\fI(true,false}\fP
Use random queueing at at all servers.  This reduces the state space.
The default is FIFO queueing.
.TP
\fBprocessor-scheduling\fR=\fIarg\fP
.RS
.TP
\fBFIFO\fP
Force all processors to use first-come, first-served scheduling.
.TP
\fBRAND\fP
Force all processors to use random scheduling for message servicing
(reduces state space)
.RE
.TP
\fBreschedule-on-async-send\fR=\fI{true,false}\fR
Force a resechedule of the processor (requeue) when an asynchronous
(send-no-reply) messages is sent if set to \fItrue\fP.
The default is to continue executing the current phase.
.TP
\fBsave-marginal-probabilities\fR=\fI{true,false}\fR
Save the marginal queue probabilites for all multiserver stations if
set to \fItrue\fP. 
.TP
\fBseverity-level\fR=\fIarg\fR
.RS
.TP
\fBall\fP
Output all error, warning and advisory messages.
.TP
\fBerror\fP
Don't output warnings or advisory messages.
.TP
\fBwarning\fP
Don't output advisory messages.
.TP
\fBadvisory\fP
Don't output warning messages.
.RE
The default is all.
.TP
\fBspex-comment\fR=\fI{true,false}\fR
Output the model comment as a header row with SPEX output.  The default is \fIfalse\fP.
.TP
\fBspex-convergence\fR=\fIarg\fR
Set the SPEX convergence value to \fIarg\fP.  
\fIArg\fP must be a number greater than 0.
SPEX convergence only applies if SPEX the convergence section is present in the input file.
It should be set to a value with \fIless\fP precision than the convergence
used by the analytic solver and far less than the expected confidence intervals expected by
the simulator.
.TP
\fBspex-header\fR=\fI{true,false}\fR
Output the header row (column names) with SPEX output.  The default is \fItrue\fP.
.TP
\fBspex-iteration-limit\fR=\fIarg\fR
Set the SPEX Iteration Limit to \fIarg\fP.  
\fIArg\fP must be a number greater than 0.
The SPEX iteration limit only applies if SPEX the convergence section is present in the input file.
.TP
\fBspex-underrelaxation\fR=\fIarg\fR
Set the SPEX underrelaxation value to \fIarg\fP.  
\fIArg\fP must be a number between 0.0 and 1.0.
The SPEX underrelaxation only applies if SPEX the convergence section is present in the input file.
.TP
\fBstop-on-message-loss\fR\fR=\fI{true,false}\fR
This pragma is used to control the operation of the solver when the
arrival rate exceeds the service rate of a server.
\fIArg\fP must be one of: 
.RS
.TP
\fBtrue\fP
Stop if messages are lost.
.TP
\fBfalse\fP
Ignore queue overflows for open arrivals and send-no-reply requests.
Since queues in petrisrvn are bounded, results may differ from
analytic solutions with no bounds.
.LP
The default is no.
.RE
.TP
\fBtask-scheduling\fR=\fIarg\fP
Set the scheduling for messsages arriving at the task to \fIarg\fP.
.RS
.TP
\fBFIFO\fP
Force all tasks to use first-come, first-served scheduling.
.TP
\fBRAND\fP
Force all tasks to use random scheduling for message servicing
(reduces state space)
.RE
.SH "RETURN VALUE"
\fBPetrisrvn\fR exits with 0 on success, 1 if the model failed to converge,
2 if the input was invalid, 4 if a command line argument was
incorrect, 8 for file read/write problems and -1 for fatal errors.  If
multiple input files are being processed, the exit code is the
bit-wise OR of the above conditions.
.SH "MODEL LIMITS"
The following table lists the acceptable parameter types for
\fBpetrisrvn\fR.  An error will be reported if an unsupported parameter
is supplied unless the value is the same as the default.
.PP
.TP 32
Phases
3
.TP 
Scheduling
FIFO, HOL, PPR, RAND
.TP
Open arrivals
yes

NEWS
* Bugs fixed in 5.31
468 	lqiolib 	bug-467.xlqn:48: error: Entry "WebServer" has a start activity defined. Activity "analyze" is a duplicate.
469 	lqiolib 	Reload json results from spex model fails. 
472 	lqiolib 	hexp(m,cv^2) not implemented.
466 	lqn2xxx 	terminating due to uncaught exception of type ActivityList::bad_internal_join: 
465 	lqn2xxx 	Generating .svg and .emf output is borked.
478 	lqns 		LQNS 6.2.28 handles correctly this model while 6.2.29 throws a lqns: domain error
222 	lqsim	 	V5 lqsim: faulty models causing aborts
264 	lqsim 		V6: lqsim regression.
453 	lqsim 		lqsim (V5) 63-semaphore failing. Priorities inverted in parasol.
481 	lqsim 		Processor utilization is NOT saved if the processor sharing is used.
476 	lqx 		C++20 [-Wtemplate-id-cdtor] 21 | ReferencedPointer<T>()
480 	miscella 	template-id not allowed for constructor in C++20 [-Wtemplate-id-cdtor]
470 	petrisrv 	The 'c' parameter has no effect.
423 	petrisrv 	Petrisrvn quorum borked.

* Bugs fixes in 5.30
208 	lqiolib 	Add --gnuplot output. 
448 	lqiolib 	Faults in open models.
456 	lqiolib 	Add drop probability for open arrivals.
458 	lqiolib 	Schema change for latest version of JMVA.
459 	lqiolib 	LQX output is not saved to a file with -ofilename 
460 	lqiolib 	Multiclass response time plotting is borked. 
384 	lqn2xxx 	Join is permitted in loop
454 	lqn2xxx 	libc++abi: terminating due to uncaught exception of type std::out_of_range: map::at: key not found
432 	lqns 		Bizarre error setting convergence value.
433 	lqns 		51-replication
452 	lqns		56-replication fails. 
449 	petrisrv 	Petrisrvn regression failures
450 	srvndiff 	srvndiff: 34-priority.lqxo:11: error: Invalid argument "-0.538461" to attribute for element <result-task>

* Bugs fixes in 5.29
441     lqiolib         Null model comment.
443     lqiolib         Export/Import the schema "description" field.
445     lqiolib         if ( va_list ) {}
384     lqn2xxx         Join is permitted in loop 
436     lqn2xxx         Model::format() -- bounding box is position sensitive. 
438     lqn2xxx         Queueing model output is not generated.
434     qnsolver        Bounds solution aborts if no whatif found in the input file.
446     qnsolver        Segmentation fault: in `LQX::ObjectPropertyReadNode::invoke() at SyntaxTree.cpp:2096:41

* Bugs fixes in 5.28
413 	Document 	Missing plainurl. 	
205 	lqiolib 	SPEX Convergence variables in XML (Json) files. 
297 	lqiolib 	Gnuplot and result enhancements
359 	lqiolib 	Expressions are not output by Spex::print_header().
411 	lqiolib 	JMVA issues.
416 	lqiolib 	Fixes for plotting population mix WhatIf output. 
422 	lqiolib 	Refactor handling of control variables. 
366 	lqn2xxx 	EXC_BAD_ACCESS Layer::transmorgrify() at layer.cc:571:17 [opt] 
428 	lqn2xxx 	lqn2fig: EXC_BAD_ACCESS libsystem_platform.dylib`_platform_strlen + 4
432 	lqns	 	Bizarre error setting convergence value. 	
433 	lqns	 	51-replication
425 	lqns	 	Refactor initPopulation. 
430 	lqns 		activity-templ.lqn:75: error: Task "Server", and-join "parB & parA" does not match and-fork "parA & parB" at line 74.
418 	lqsim 		lqsim crashes running -t... 	
429 	lqsim 		result.cc:308:12: warning: 'vsprintf' is deprecated: 
356 	miscella 	rpm build issues.
415 	petrisrv 	Signal SIGSEGV in LQIO::DOM::Phase::getServiceTime () at dom_phase.cpp:88 
421 	petrisrv 	SPEX Convergence with PetriSRVN 
423 	petrisrv 	Petrisrvn quorum borked. 
424 	petrisrv 	09-sanity.xlqn 
343 	qnsolver 	Improve compatibility with Java Modelling Tools 
412 	qnsolver 	Save JMVA WhatIf results.
417 	qnsolver 	Bounds solution issues. 
419 	qnsolver 	model.cc:(.text+0x18ff): undefined reference to `typeinfo for QNIO::JMVA_Document'

* Bugs fixed in 5.27
192 	libmva 	 	MVA::marginalProbabilities2 mis-dimensioned. 
401 	libmva 		libc++abi: terminating with uncaught exception of type std::invalid_argument: default-output
403 	libmva 	 	43-activities.lqnx: Floating point exception: mva.cc 541: Inexact result, Invalid operation
406 	libmva 	 	Refactor fpgoop.
383 	lqiolib 	<pre> precedences can accept more than one activity in XML (JSON?) input. 
393 	lqiolib 	Save (and print) marginal queue probabilities.
397 	lqiolib 	Add qnap2 parser for BCMP models.
398 	lqiolib 	Refactor LQX math handing from ExternalVariables to LQX Expressions.
399 	lqiolib 	WhatIf construction is bogus.
411 	lqiolib 	JMVA issues. 
386 	lqn2csv 	getopt_long headaches. 
382 	lqn2xxx 	lqn2ps -Oout doesn't print processors. 
390 	lqn2xxx 	lqn2xxx: command not found. 
400 	lqn2xxx 	Converting from json to lqn fails on Windows.
373 	lqns 	 	Processor waiting times at Rolia multiserver make no sense. 
380 	lqns 	 	lqns hangs with Conway multiserver. 
405 	lqns 	 	Optimizer bug: option.h
407 	lqns 	 	--srvn-layering crashes in V6 interlocking code
391 	lqsim 		lqsim/regression/Makefile.am, 5*-replication. 
356 	miscella 	rpm build issues.
402 	miscella 	Mac M1 issues.
385 	petrisrv 	V5 petrisrvn: faulty models causing aborts.
392 	petrisrv 	Calls are not being constructed
395 	petrisrv 	V5 petrisrvn: faulty models causing aborts or failing.
409 	petrisrv 	Petrisrvn fails when SPEX/LQX is used to alter the arrival rate to an entry. 
387 	srvndiff 	File glob not working. 

petrisrvn/petrisrvn.cc
/************************************************************************/
/* Copyright the Real-Time and Distributed Systems Group,		*/
/* Department of Systems and Computer Engineering,			*/
/* Carleton University, Ottawa, Ontario, Canada. K1S 5B6		*/
/* 									*/
/* September 1991.							*/
/* August 2003.								*/
/************************************************************************/

/*
 * $Id: petrisrvn.cc 17507 2024-12-04 01:52:14Z greg $
 *
 * Generate a Petri-net from an SRVN description.
 *
 */

#include "petrisrvn.h"
#include <iostream>
#include <cstdlib>
#include <cstring>
#include <cassert>
#include <errno.h>
#if HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#if HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <signal.h>
#if HAVE_FENV_H
#include <fenv.h>
#elif HAVE_IEEEFP_H && !defined(MSDOS)
#include <ieeefp.h>
#endif
#if HAVE_FLOAT_H
#include <float.h>
#endif
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <libgen.h>
#include <lqio/commandline.h>
#include <lqio/dom_document.h>
#include <lqio/error.h>
#include <lqio/filename.h>
#include <lqio/glblerr.h>
#include <lqio/srvn_spex.h>
#if !HAVE_GETSUBOPT
#include <lqio/getsbopt.h>
#endif
#include <wspnlib/global.h>
#include <wspnlib/wspn.h>
#include "errmsg.h"
#include "model.h"
#include "pragma.h"

static bool copyright_flag	= false; /* Print copyright notice	*/

bool debug_flag                 = false; /* Debugging flag set?         */
bool keep_flag                  = false; /* Keep result files?          */
bool no_execute_flag            = false; /* Don't solve model if true   */
bool reload_net_flag            = false; /* Reload results and print.   */
bool rtf_flag                   = false; /* Parsable output desired?    */
bool trace_flag			= false; /* Output execution of greatspn*/
bool uncondition_flag           = false; /* Uncondition in-service      */
bool verbose_flag               = false; /* Verbose text output?        */

bool distinguish_join_customers = true;	 /* unique cust at join for mult*/
bool simplify_network		= false; /* Delete single place procs.  */

double	x_scaling		= 1.0;	 /* Auto-squish if val == 0.	*/

static const char * net_dir_name	= "nets";
static void my_handler (int);

/*
 * Command options.
 */

static const char * opts = "dHI:jkm:no:pP:Q:rRtvVwxz:";

#if HAVE_GETOPT_LONG
static const struct option longopts[] =
    /* name */ /* has arg */ /*flag */ /* val */
{
    { "debug",              no_argument,        0, 'd' },
    { "help",               no_argument,        0, 'H' },
    { "input-format",	    required_argument,  0, 'i' },
    { "json",		    no_argument,	0, 'j' },
    { "keep-net",           no_argument,        0, 'k' },
    { "monitor-file",       required_argument,  0, 'm' },
    { "no-execute",         no_argument,        0, 'n' },
    { "output",             required_argument,  0, 'o' },
    { "parseable",          no_argument,        0, 'p' },
    { "pragma",             required_argument,  0, 'P' },
    { "queue-size",         required_argument,	0, 'Q' },
    { "rtf",	            no_argument,        0, 'r' },
    { "reload-net",	    no_argument,	0, 'R' },
    { "trace-greatspn",     no_argument,        0, 't' },
    { "verbose",            no_argument,        0, 'v' },
    { "version",            no_argument,        0, 'V' },
    { "no-warnings",        no_argument,        0, 'w' },
    { "xml",                no_argument,        0, 'x' },
    { "disjoint-customers", no_argument,	0, 256+'d' },
    { "no-header",	    no_argument,	0, 256+'h' },
    { "print-comment",	    no_argument,	0, 256+'c' },
    { "overtaking",	    optional_argument,	0, 256+'O' },
    { "random-processors",  no_argument,	0, 256+'p' },
    { "random-queueing",    no_argument,	0, 256+'q' },
    { "reload-lqx",	    no_argument,        0, 256+'r' },
    { "restart",	    no_argument,	0, 256+'R' },
    { "simplify",	    no_argument,	0, 256+'s' },
    { "spacing",	    required_argument,	0, 256+'S' },
    { "debug-lqx",          no_argument,        0, 256+'l' },
    { "debug-xml",          no_argument,        0, 256+'x' },
    { nullptr, 0, 0, 0 }
};
#else
const struct option * longopts = nullptr;
#endif

static const char * opthelp[]  = {
    /* "debug"		    */    "Enable debug code.",
    /* "help"		    */    "Show this help.",
    /* "input-format"	    */    "Force input format to ARG.  Arg is either 'lqn' or 'xml'.",
    /* "json"		    */    "Output results in JSON format.",
    /* "keep-net"	    */    "Keep the files generated by GreatSPN.",
    /* "monitor-file"	    */    "Output results from GreatSPN to FILE.",
    /* "no-execute"	    */    "Build the Petri Net, but do not solve.",
    /* "output"		    */    "Redirect solver ouptut to FILE.",
    /* "parseable"	    */    "Generate parseable (.p) output.",
    /* "pragma"		    */    "Set solver options.",
    /* "rtf"		    */    "Output results in Rich Text Format.",
    /* "reload-net"	    */    "Reload results from GreatSPN solution.",
    /* "trace-greatspn"     */    "Show output from the execution of the GreatSPN solver.",
    /* "verbose"	    */    "Output on standard error the progress of the simulation.",
    /* "version"	    */    "Print the version of the simulator.",
    /* "no-warnings"	    */    "Do not output warning messages.",
    /* "xml"		    */    "Output results in XML format.",
    /* "disjoint-customers" */    "Create copies for reference tasks (increases state space).",
    /* "no-header"	    */	  "Do not output the variable name header on SPEX results.",
    /* "print-comment"	    */    "Add the model comment as the first line of output when running with SPEX input.",
    /* "overtaking"	    */	  "Find in-service and overtaking probabilities (increases state space). ARG=condition",
    /* "queue-limit    	    */	  "Set the maximum queue size (open model).",
    /* "random-processors"  */	  "Use random queueing at all processors (reduces state space).",
    /* "random-queueing"    */    "Use random queueing at all tasks and processors (reduces state space).",
    /* "reload-lqx"	    */    "Run the LQX program, but re-use the results from a previous invocation.",
    /* "restart"	    */    "Reuse existing results where available, but solve any unsolved models.",
    /* "simplify"	    */    "Remove redundant parts of the net such as single place processors.",
    /* "spacing"	    */	  "Set the spacing between places and transitions to ARG.",
    /* "debug-lqx"	    */    "Output debugging information while parsing LQX input.",
    /* "debug-xml"	    */    "Output debugging information while parsing XML input.",
    nullptr
};


std::regex * inservice_match_pattern	= nullptr;

FILE * stddbg;				/* debugging output goes here.	*/

static void usage( void );
 
int
main(int argc, char *argv[])
{
    std::string output_file = "";
    LQIO::DOM::Document::InputFormat input_format = LQIO::DOM::Document::InputFormat::AUTOMATIC;
    LQIO::DOM::Document::OutputFormat output_format = LQIO::DOM::Document::OutputFormat::DEFAULT;
    LQIO::CommandLine command_line( longopts );
    LQIO::DOM::Pragma pragmas;
    Model::solve_using solve_function = &Model::compute;

    extern char *optarg;
    extern int optind;

    int status  = NORMAL_TERMINATION;

    LQIO::io_vars.init( VERSION, basename( argv[0] ), LQIO::severity_action );
    std::copy( local_error_messages.begin(), local_error_messages.end(), std::inserter( LQIO::error_messages, LQIO::error_messages.begin() ) );

    command_line = LQIO::io_vars.lq_toolname;

    /* Check for non-empty non-regular file "empty" -- used by GreatSPN */
#if HAVE_SYS_STAT_H
    struct stat statbuf;
    if ( stat( "empty", &statbuf ) == 0 && ((statbuf.st_mode & S_IFMT) != S_IFREG || statbuf.st_size != 0) ) {
	fprintf( stderr, "%s: \"empty\", used by GreatSPN, is present in the working directory\n", LQIO::io_vars.toolname() );
	exit( 1 );
    }
#endif

    stddbg   = stdout;

#if 0
    inter_proc_delay = 0.0;
    comm_delay_flag  = false;
#endif

#if defined(HAVE_FENV_H) && defined(HAVE_FEENABLEEXCEPT)
    feenableexcept( FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW );
#elif  defined(HAVE_IEEEFP_H) && defined(HAVE_FPSETMASK)
    fpsetmask( FP_X_INV | FP_X_DZ | FP_X_OFL );
#endif
    signal(SIGFPE, my_handler);

    pragmas.insert( getenv( "PETRISRVN_PRAGMAS" ) );

    LQIO::DOM::DocumentObject::setSeverity(LQIO::ERR_NO_QUANTUM_SCHEDULING, LQIO::error_severity::WARNING );	// Don't care for petrisrvn.

    for ( ;; ) {
#if HAVE_GETOPT_LONG
	const int c = getopt_long( argc, argv, opts, longopts, NULL );
#else
	const int c = getopt( argc, argv, opts );
#endif
	if ( c == EOF) break;

	command_line.append( c, optarg );

	switch ( c ) {
	case 'c'+256:
	    LQIO::Spex::__print_comment = true;
	    pragmas.insert(LQIO::DOM::Pragma::_spex_comment_,"true");
	    break;

	case 'd':
	    debug_flag = true;
	    keep_flag  = true;
	    break;

	case 256+'d':
	  //	    distinguish_join_customers = true;			/* wrong one! */
	    pragmas.insert(LQIO::DOM::Pragma::_disjoint_customers_);
	    break;

	case 'k':
	    keep_flag = true;
	    break;

	case 256+'h':
	    pragmas.insert(LQIO::DOM::Pragma::_spex_header_,LQIO::DOM::Pragma::_false_);
	    LQIO::Spex::__no_header = true;
	    break;

	case 'H':
	    usage();
	    exit( 0 );

	case 'j':
	    output_format = LQIO::DOM::Document::OutputFormat::JSON;
	    break;

	case 'I':
	    if ( strcasecmp( optarg, "xml" ) == 0 ) {
		input_format = LQIO::DOM::Document::InputFormat::XML;
	    } else if ( strcasecmp( optarg, "json" ) == 0 ) {
		input_format = LQIO::DOM::Document::InputFormat::JSON;
	    } else if ( strcasecmp( optarg, "lqn" ) == 0 ) {
		input_format = LQIO::DOM::Document::InputFormat::LQN;
	    } else {
		fprintf( stderr, "%s: invalid argument to -I -- %s\n", LQIO::io_vars.toolname(), optarg );
	    }
	    break;

	case 256+'l':
	    LQIO::DOM::Document::lqx_parser_trace(stderr);
	    break;

	case 'm':
	    trace_flag = true;
	    if ( strcmp( optarg, "-" ) == 0 ) {
		stddbg = stdout;
	    } else if ( !(stddbg = fopen( optarg, "w" )) ) {
		(void) fprintf( stderr, "%s: cannot open ", LQIO::io_vars.toolname() );
		perror( optarg );
		(void) exit( FILEIO_ERROR );
	    }
	    break;

	case 'n':
	    no_execute_flag = true;
	    keep_flag = true;
	    break;

	case 'o':
	    output_file = optarg;
	    break;

	case 'p':
	    output_format = LQIO::DOM::Document::OutputFormat::PARSEABLE;
	    break;

	case 'P':
	    if ( !pragmas.insert( optarg ) ) {
		Pragma::usage( std::cerr );
		exit( INVALID_ARGUMENT );
	    }
	    break;

	case 256+'p':
	    pragmas.insert(LQIO::DOM::Pragma::_processor_scheduling_,LQIO::SCHEDULE::RAND);
	    break;

	case 256+'q':
	    pragmas.insert(LQIO::DOM::Pragma::_force_random_queueing_);
	    break;

	case 256+'s':
	    simplify_network = true;
	    break;

	case 'Q':
	    pragmas.insert(LQIO::DOM::Pragma::_queue_size_,optarg);
	    break;

	case 'r':
	    rtf_flag = true;
	    break;

	case 'R':
	    reload_net_flag = true;
	    no_execute_flag = false;	/* Save results */
	    keep_flag = true;
	    break;

	case 256+'r':
	    solve_function = &Model::reload;
	    break;

	case 256+'R':
	    solve_function = &Model::restart;
	    break;

	case 256+'S':
	    x_scaling = 1.0;
	    if ( optarg == nullptr || sscanf( optarg, "%lg", &x_scaling ) != 1 || x_scaling < 0 ) {
		(void) fprintf( stderr, "%s: x_scaling=%s is invalid, choose value > 0\n",
				LQIO::io_vars.toolname(), optarg != nullptr ? optarg : "" );
		(void) exit( INVALID_ARGUMENT );
	    }
	    break;

	case 't':
	    trace_flag = true;
	    keep_flag = true;
	    break;

	case 'v':
	    verbose_flag = true;
	    LQIO::Spex::__verbose = true;
	    break;

	case 'V':
	    copyright_flag = true;
	    break;

	case 'w':
	    pragmas.insert(LQIO::DOM::Pragma::_severity_level_,LQIO::DOM::Pragma::_run_time_);
	    break;

	case 'x':
	    output_format = LQIO::DOM::Document::OutputFormat::XML;
	    break;

        case (256+'x'):
	    LQIO::DOM::Document::__debugXML = true;
	    break;

	case 256+'O':
	    if ( optarg == nullptr ) {
		inservice_match_pattern = new std::regex( ".*" );
	    } else {
		inservice_match_pattern = new std::regex( optarg );
	    }
#if 0
		case UNCONDITIONAL_PROB:
		    uncondition_flag = true;
		    break;

#endif
	    break;

	default:
	    usage();
	    exit( INVALID_ARGUMENT );
	}
    }
    LQIO::io_vars.lq_command_line = command_line.c_str();

    if ( copyright_flag ) {
	(void) fprintf( stdout, "\nStochastic Rendezvous Petri Network Analyser, Version %s\n\n", VERSION );
	(void) fprintf( stdout, "  Copyright the Real-Time and Distributed Systems Group,\n" );
	(void) fprintf( stdout, "  Department of Systems and Computer Engineering,\n" );
	(void) fprintf( stdout, "  Carleton University, Ottawa, Ontario, Canada. K1S 5B6\n\n");
    }

    /* Quick check.  -zin-service requires that the customers are differentiated. */

    if ( ( inservice_match_pattern != nullptr ) && !Pragma::__pragmas->disjoint_customers() ) {
	if ( LQIO::io_vars.severity_level == LQIO::error_severity::ERROR || LQIO::io_vars.severity_level == LQIO::error_severity::ALL ) {
	    (void) fprintf( stdout, "%s: --overtaking is incompatible with multiple customer clients\n", LQIO::io_vars.toolname() );
	    (void) fprintf( stdout, "\t--disjoint-customers assumed\n" );
	}
	Pragma::__pragmas->set_disjoint_customers( "false" );
    }

    /* create net_dir_name directory */

    make_net_dir( net_dir_name );

    /* Process all command line arguments.  If none specified, then	*/
    /* input is assumed to come in from stdin.			*/

    if ( optind == argc ) {

	status |= Model::solve( solve_function, "-", input_format, output_file, output_format, pragmas );

    } else {
	unsigned int file_count = argc - optind;			/* Number of files on cmd line	*/

	if ( file_count > 1 && LQIO::Filename::isFileName( output_file ) && !std::filesystem::is_directory( output_file ) ) {
	    (void) fprintf( stderr, "%s: Too many input files specified with -o <file> option.\n", LQIO::io_vars.toolname() );
	    exit( INVALID_ARGUMENT );
	}

	for ( ; optind < argc; ++optind ) {

	    if ( file_count > 1 ) {
		(void) printf( "%s:\n", argv[optind] );
	    }

	    status |= Model::solve( solve_function, argv[optind], input_format, output_file, output_format, pragmas );
	}

    }

    unlink( "empty" );		/* Clean up after GreatSPN. */

    return status;	 	/* 0 is a successful exit */
}

static void
usage (void)
{
    (void) fprintf( stderr, "Usage: %s ", LQIO::io_vars.toolname());

#if HAVE_GETOPT_LONG
    fprintf( stderr, " [option] [file ...]\n\n" );
    fprintf( stderr, "Options:\n" );
    const char ** p = opthelp;
    for ( const struct option *o = longopts; (o->name || o->val) && *p; ++o, ++p ) {
	std::string s;
	if ( o->name ) {
	    s = "--";
	    s += o->name;
	    switch ( o->val ) {
	    case 'o':
	    case 'm': s += "=<file>"; break;
	    }
	} else {
	    s = " ";
	}
	if ( isascii(o->val) && isgraph(o->val) ) {
	    fprintf( stderr, " -%c, ", static_cast<char>(o->val) );
	} else {
	    fprintf( stderr, "     " );
	}
	fprintf( stderr, "%-24s %s\n", s.c_str(), *p );
    }
#else
    for ( char * s = opts; *s; ++s ) {
	if ( *(s+1) == ':' ) {
	    ++s;
	} else {
	    fputc( *s, stderr );
	}
    }
    std::cerr << ']';
    for ( char * s = opts; *s; ++s ) {
	if ( *(s+1) == ':' ) {
	    fprintf( stderr, " [-%c", *s );
	    switch ( *s ) {
	    default: fprintf( stderr, "file" ); break;
	    }
	    fputf( ']', stderr );
	    ++s;
	}
    }
    fprintf( stderr, " [file ...]\n" );
#endif
}


/*
 * signal handler for fp errors.
 */

static void
my_handler (int sig )
{
    abort();
}
